caikit/runtime/service_generation/data_stream_source.py

```diff
@@ -255,53 +202,354 @@ def _get_resolved_source_path(input_path: str) -> str:
+
+        # Determine which of the value types is set
+        set_field = None
+        for field_name in self.get_proto_class().DESCRIPTOR.fields_by_name:
+            if getattr(self, field_name) is not None:
+                error.value_check(
+                    "<COR80421785E>",
+                    set_field is None,
+                    "Found DataStreamSource with multiple sources set: {} and {}",
+                    set_field,
+                    field_name,
+                )
+                error.value_check(
+                    "<COR80420785E>",
+                    field_name in self.name_to_plugin_map,
+                    "no data stream plugin found for field: {}",
+                    field_name,
+                )
+                set_field = field_name
+
+        # If no field is set, return an empty DataStream
+        if set_field is None:
+            log.debug3("Returning empty data stream")
+            return DataStream.from_iterable([])
+
+        # Get the correct plugin, and pass it the source field + the element
+        # type to serialize to
+        plugin = self.name_to_plugin_map[set_field]
+        return plugin.to_data_stream(getattr(self, set_field), self.ELEMENT_TYPE)
+
+
+## make_data_stream_source #####################################################
+
+
+def make_data_stream_source(
+    data_element_type: type,
+    plugin_factory: DataStreamPluginFactory = PluginFactory,
+    plugins_config: Optional[aconfig.Config] = None,
+) -> Type[DataBase]:
     """Dynamically create a data stream source message type that supports
     pulling an iterable of the given type from all valid data stream sources
     """
     log.debug2("Looking for DataStreamSource[%s]", data_element_type)
     if data_element_type not in _DATA_STREAM_SOURCE_TYPES:
         cls_name = _make_data_stream_source_type_name(data_element_type)
+        package = get_runtime_service_package()
+
         log.debug("Creating DataStreamSource[%s] -> %s", data_element_type, cls_name)
 
-        # Set up the "sub class." In python, this is the same as creating a
-        # standalone class with a __qualname__ that nests it under a parent
-        # class. We do this outside the declaration of the parent class so that
-        # this class can be referenced within the Union annotation for the
-        # outer class itself. This class needs to be created dynamically because
-        # it encapsulates type information about the elements of the data stream.
-        package = get_runtime_service_package()
-        JsonData = make_dataobject(
-            package=package,
-            proto_name=f"{cls_name}JsonData",
-            name="JsonData",
-            attrs={"__qualname__": f"{cls_name}.JsonData"},
-            annotations={"data": List[data_element_type]},
+        # Get the required plugins
+        plugins = plugin_factory.get_plugins(plugins_config)
+
+        # Make sure there are no field name duplicates
+        plug_to_name = {
```

Comment:
This work to check for duplicates isn't free, but it should all be boot-time work, so I'm ok not putting this behind an enablement flag to turn of for performance